### ISSUE TITLE
[all] Whitelisting FrostWire on Trackers

### ISSUE THREAD

**Issue by gubatron** [frostwire/frostwire#786]
Per [community feedback](https://www.reddit.com/r/torrents/comments/956mtu/frostwire_team_asks_for_feedback/e3qgtxh/), FrostWire isn't white listed among private tracker communities.

1. Start announcing the client under torrent client identifier `FW`
2. Submit a patch to [bep_0020](http://www.bittorrent.org/beps/bep_0020.html) once we've rolled out our torrent client identifier
3. Contact /r/trackers and staffs on IRC to submit FrostWire for testing so they can see it reports upload/download statistics accordingly, adjust whatever settings they might require.

Thanks brickfrog2

**Comment by aldenml** [comment]
Given that we are essentially a `libtorrent` wrapper for the bittorrent part, what practical advantages this give to us? I have no problem with this change, but I would like to know

**Comment by gubatron** [comment]
Thinking about it more, an easier way to put it is, right now I bet you can't use FrostWire on most private trackers and that's unacceptable since we know we have a world class bittorrent client here.
It needs to be whitelisted

**Comment by popcornenthusiast** [comment]
I think FrostWire only will be whitelisted when it offers extensible IP filtering capabilities. So, without #702 and even #466 (cause, to be fair, IP filtering alone does almost nothing), it won't be accepted in private trackers or simply by privacy conscious users.

**Comment by gubatron** [comment]
Update:
First step done. [Advertise the client's user agent and fingerprint](https://github.com/frostwire/frostwire/commit/535c71684bc0ed4fff6ec1f9d327238e168941ae) so tracker operators can identify, evaluate and whitelist.

@popcornenthusiast  Next up, I'll try to take a jab at IP Filtering, and will look again at findspies.

Q: do you know any IRC channels of private trackers on freenode so I can start a conversation to see requirements, testing, getting an account.

**Comment by gubatron** [comment]
thanks for that, I do know Ernesto from TorrentFreak, maybe he has contacts

**Comment by gubatron** [comment]
Update: I've started working on the `IP Filter` tool on [this branch](https://github.com/frostwire/frostwire/commits/ip-filter)

**Comment by popcornenthusiast** [comment]
> Update: I've started working on the IP Filter tool on this branch

Great!

And here to expand on something a [reddit user said](https://www.reddit.com/r/torrents/comments/956mtu/frostwire_team_asks_for_feedback/e3qskqs/) and on IP filtering:

> I see no bittorrent option for a whitelist, or blacklist. No Protocol encryption either. Is this because of the advertisements for paid VPN?

This is how protocol encryption works in uTorrent and why it is so useful: https://whirlpool.net.au/wiki/utorrent_encryption

Aside from preventing BitTorrent traffic throttling/blocking, there is also one really interesting thing about it: A lot of BitTorrent watchers/fake peers, to save bandwidth, only use non-encrypted connections, so blocking non-encrypted seeds/peers connections also blocks them.

> First step done. Advertise the client's user agent and fingerprint so tracker operators can identify, evaluate and whitelist.

On this same matter, blocking peers that don't properly identify themselves may be of good use, as it is also common among watchers/fake peers.

**Comment by aldenml** [comment]
I want to clarify that that "encryption" is more an obfuscation using a very insecure RC4. Today, this PE does not prevent traffic shaping analysis by the ISP. If watchers/fake peers don't use it, I don't know, but limiting yourself to only establish PE connections will hurt your P2P performance for sure.

In any case, it's just a number in the settings that we need to expose with a checkbox in the configuration. For development reference:

```cpp
allowed_enc_level

// the encryption levels, to be used with
// settings_pack::allowed_enc_level.
enum enc_level
{
    // use only plaintext encryption
    pe_plaintext = 1,
    // use only rc4 encryption
    pe_rc4 = 2,
    // allow both
    pe_both = 3
};
```

**Comment by popcornenthusiast** [comment]
@aldenml Yeah, it's old and insecure, but it does the job of preventing throttling, so, still useful.

> If watchers/fake peers don't use it, I don't know,

Try with some new copyrighted protected content, and you'll see a lot of "hit-and-run" peers appearing, and that a lot of them don't use encryption.

> but limiting yourself to only establish PE connections will hurt your P2P performance for sure.

Definitely possible. Never had a problem with it, though.

> In any case, it's just a number in the settings that we need to expose with a checkbox in the configuration. For development reference:

That seems kinda strict? Like I referenced, in uTorrent you can just: allow non-encrypted incoming connections AND don't use PE for outgoing connections; allow non-encrypted incoming connections AND try to use PE for outgoing connections; allow non-encrypted incoming connections AND only make PE outgoing connections; block non-encrypted incoming connections AND don't use PE for outgoing connections; block non-encrypted incoming connections AND try to use PE for outgoing connections; block non-encrypted incoming connections AND only make PE outgoing connections. So, yeah, a lot more than just "use or don't use PE".

That said, having PE activated by default (but not forcing it) for outgoing connections will, at some point, lead to being able to block non-encrypted incoming connections by default. Until that day comes, just let users decide if they need to only accept PE incoming connections.

**Comment by aldenml** [comment]
interesting all the options uTorrent has related to PE, I'm not aware of that in `libtorrent` (which is the one we wrap), but I will look at it in the source code
